File: chris_cunningham/day02.py
import re

r = re.compile(r"(\d+)-(\d+) (\w): (\w+)")


def part_a_test(matches: re.Match) -> bool:
    low, high, char, pw = matches.groups()
    return int(low) <= pw.count(char) <= int(high)


def part_b_test(matches: re.Match) -> bool:
    low, high, char, pw = matches.groups()
    c = (pw[int(low)-1] == char, pw[int(high)-1] == char)
    return sum(1 for i in c if i) == 1


with open("inputs/day02.txt", 'r') as f:
    data = [r.match(item) for item in f.readlines()]

    part_a = sum(1 for i in data if part_a_test(i))
    print(f"part_a: {part_a}")

    part_b = sum(1 for i in data if part_b_test(i))
    print(f"part_b: {part_b}")

File: chris_cunningham/day02.rs
use std::fs;
use regex::{Regex, Captures};


fn part_a_test(matches: &Captures) -> bool {
    let low = matches.get(1).unwrap().as_str().parse::<usize>().unwrap();
    let high = matches.get(2).unwrap().as_str().parse::<usize>().unwrap();
    let char = matches.get(3).unwrap().as_str();
    let pw = matches.get(4).unwrap().as_str();
    let c = pw.matches(char).count();
    return c >= low && c <= high;
}

fn part_b_test(matches: &Captures) -> bool {
    let low = matches.get(1).unwrap().as_str().parse::<usize>().unwrap();
    let high = matches.get(2).unwrap().as_str().parse::<usize>().unwrap();
    let char = matches.get(3).unwrap().as_str().chars().nth(0).unwrap();
    let pw = matches.get(4).unwrap().as_str();
    return (pw.chars().nth(low - 1).unwrap() == char) ^ (pw.chars().nth(high - 1).unwrap() == char);
}

pub fn main() {
    let r: Regex = Regex::new(r"(\d+)-(\d+) (\w): (\w+)").unwrap();

    let data = fs::read_to_string("inputs/day02.txt").unwrap();
    let data: Vec<Captures> = data
        .lines()
        .map(|s| r.captures(s).unwrap())
        .collect();

    let part_a = data.iter()
        .filter(|x| part_a_test(x))
        .count();
    println!("part a: {}", part_a);

    let part_b = data.iter()
        .filter(|x| part_b_test(x))
        .count();
    println!("part b: {}", part_b);
}
